### PR TITLE
Compcert sumhash integration

### PR DESCRIPTION
**Pull request by algonathan**
## Summary
I've changed compactcert to support genericDigest and to use sumhash throughout the package, and it'

### CODE REVIEW

File: crypto/digest.go [hunk 1]
```diff
@@ -16,17 +16,36 @@
...
-	var cpy [32]byte
+func (d GenericDigest) To32Byte() [Sha512_256Size]byte {
+	var cpy [Sha512_256Size]byte
 	copy(cpy[:], d)
 	return cpy
 
 }
 
+// ToSumhashDigest converts the generic hash into a byte array of size: SumhashDigestSize
+func (d GenericDigest) ToSumhashDigest() [SumhashDigestSize]byte {
```
Comment (algoidan) [inline crypto/digest.go:34]: this is unused

File: test/e2e-go/features/compactcert/compactcert_test.go [hunk 2]
```diff
@@ -134,7 +133,7 @@ func TestCompactCerts(t *testing.T) {
 			err = protocol.Decode(nextCertBlockRaw, &nextCertBlockDecoded)
 			r.NoError(err)
 
-			var votersRoot crypto.Digest
+			var votersRoot = make([]byte, len(lastCertBlock.CompactCertVoters))
```
Comment (algoidan) [inline test/e2e-go/features/compactcert/compactcert_test.go:136]: we should create a CompactCert Hash size

File: crypto/compactcert/hash.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
```
Comment (algoidan) [inline crypto/compactcert/hash.go:1]: can it be changed to consthash

File: crypto/compactcert/verifier.go [hunk 4]
```diff
@@ -48,28 +48,33 @@ func (v *Verifier) Verify(c *Cert) error {
...
 	}
 
-	hsh, err := crypto.HashFactory{HashType: crypto.Sha512_256}.NewHash()
+	// Verify all of the reveals
+	sHash, err := c.SigProofs.HashFactory.NewHash()
+	if err != nil {
+		return err
+	}
+
+	pHash, err := c.PartProofs.HashFactory.NewHash()
```
Comment (algoidan) [inline crypto/compactcert/verifier.go:57]: is it possible to extract the correct hash function inside the verification of the tree

File: crypto/merklearray/merkle.go [hunk 5]
```diff
@@ -214,26 +214,47 @@ func (tree *Tree) buildNextLayer() {
...
-func Verify(root crypto.GenericDigest, elems map[uint64]crypto.GenericDigest, proof *Proof) error {
+func Verify(root crypto.GenericDigest, elems map[uint64]crypto.Hashable, proof *Proof) error {
+
+	if proof == nil {
+		return fmt.Errorf("proof should not be nil")
+	}
+
 	if len(elems) == 0 {
-		if proof != nil && len(proof.Path) != 0 {
+		if proof.Path != nil && len(proof.Path) != 0 {
```
Comment (algonathan) [inline crypto/merklearray/merkle.go:241]: you can drop the check for nil and just check length here.

File: crypto/merklearray/merkle.go [hunk 6]
```diff
@@ -214,26 +214,47 @@ func (tree *Tree) buildNextLayer() {
...
 			return fmt.Errorf("non-empty proof for empty set of elements")
 		}
 		return nil
 	}
 
-	pl := buildPartialLayer(elems)
+	var hashedLeafs map[uint64]crypto.GenericDigest
```
Comment (algonathan) [inline crypto/merklearray/merkle.go:247]:
```
var hashedLeafs map[uint64]crypto.GenericDigest
	var err error
	if hashedLeafs, err = hashLeafs(elems, proof); err != nil {
		return err
	}
```

change into: 

```
hashedLeafs, err := hashLeafs(elems, proof); 
	if  err != nil {
		return err
	}
```
?

File: crypto/merklearray/merkle.go [hunk 7]
```diff
@@ -214,26 +214,47 @@ func (tree *Tree) buildNextLayer() {
 	tree.Levels = append(tree.Levels, newLayer)
 }
 
+func hashLeafs(elems map[uint64]crypto.Hashable, proof *Proof) (map[uint64]crypto.GenericDigest, error) {
```
Comment (algonathan) [inline crypto/merklearray/merkle.go:217]: maybe change the def of the function to:

```
func hashLeafs(elems map[uint64]crypto.Hashable, factory crypto.HashFactory) (map[uint64]crypto.GenericDigest, error) {
```
?

File: crypto/merklekeystore/keystore.go [hunk 8]
```diff
@@ -145,7 +145,7 @@ func New(firstValid, lastValid, interval uint64, sigAlgoType crypto.AlgorithmTyp
 		Interval:            interval,
 		mu:                  deadlock.RWMutex{},
 	}
-	tree, err := merklearray.Build(s, crypto.HashFactory{HashType: crypto.Sumhash})
+	tree, err := merklearray.Build(s, crypto.HashFactory{HashType: crypto.Sha512_256})
```
Comment (algonathan) [inline crypto/merklekeystore/keystore.go:148]: maybe add const to be used throughout this package?

**Pull request merged**